### ISSUE TITLE
503 when the ingress is deployed after the controller

### ISSUE THREAD

**Issue by Nainterceptor** [haproxytech/kubernetes-ingress#209]
Hello,

First, thanks for this opensource product, and thanks again to provide k8s controller + helm chart.
I can use real names, because everything in our project is (or will be) opensource too. (application, charts, terraform).

So, I've a behavior that I don't understood. My latest conclusion is that it's a bug from kubernetes-ingress : 

I've an application deployed on namespace _stargate-staging_ (thanks to [this nodejs chart](https://gitlab.com/fabnum-minarm/igloo/helm-charts/-/tree/master/stable/nodejs)). My controller is in namespace `gateway`.

If I deploy (thanks to helm) my application before haproxy, it's working (and following upgrades, thanks to `helm upgrade --install` are OK)
If I deploy my application after haproxy, it's not working.
If I delete the controller pod (to force a new one to be spawned), it's working...

Any idea ? 
Bonus question : How to troubleshot ? :/

Thanks !

Configurations : 
> `kubectl get ingress -o yaml`
```yaml
apiVersion: v1
items:
- apiVersion: extensions/v1beta1
  kind: Ingress
  metadata:
    annotations:
      acme.cert-manager.io/http01-edit-in-place: "true"
      cert-manager.io/cluster-issuer: letsencrypt-prod
      kubernetes.io/ingress.class: haproxy
      meta.helm.sh/release-name: backend
      meta.helm.sh/release-namespace: stargate-staging
    labels:
      app.kubernetes.io/instance: backend
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: nodejs
      helm.sh/chart: nodejs-0.0.2
    name: backend-nodejs
    namespace: stargate-staging
  spec:
    rules:
    - host: <host>
      http:
        paths:
        - backend:
            serviceName: backend-nodejs
            servicePort: 80
          pathType: ImplementationSpecific
    tls:
    - hosts:
      - <host>
      secretName: backend-nodejs-tls
  status:
    loadBalancer: {}
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

**Comment by Mo3m3n** [comment]
Hi @Nainterceptor 
With helm you can set logging to debug: https://github.com/haproxytech/helm-charts/blob/master/kubernetes-ingress/values.yaml#L170
So you can start by sending the logs in the case where it is not working (installing the application after haproxy controller) via `kubectl logs $POD`
Also looking to the HAProxy config would be useful: 
`kubectl exec -it $POD cat /etc/haproxy/haproxy.cfg`

**Comment by Dr4s1l** [comment]
Following this issue, by looking at processes lauched by the ingress controller pod, we have seen 2 haproxy processes simultanouesly.

```
    1 root      0:00 /dumb-init -- /start.sh --default-ssl-certificate=gateway/haproxy-kubernetes-ingress-default-cert --configmap=gateway/haproxy-kubernetes-ingress --default-backend-service=gateway/haproxy-kubernetes-ingress-default-backend --ingress.class=haproxy --log=info --ignore-ingress-without-class
    6 root     49:19 /haproxy-ingress-controller --default-ssl-certificate=gateway/haproxy-kubernetes-ingress-default-cert --configmap=gateway/haproxy-kubernetes-ingress --default-backend-service=gateway/haproxy-kubernetes-ingress-default-backend --ingress.class=haproxy --log=info --ignore-ingress-without-class
   25 root      0:00 [haproxy]
   29 root      0:09 haproxy -W -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -sf 357 -x /var/run/haproxy-runtime-api.sock
  374 root      4:20 haproxy -W -f /etc/haproxy/haproxy.cfg -p /var/run/haproxy.pid -sf 357 -x /var/run/haproxy-runtime-api.sock
```

Is this normal, we see in options that it kill a process PID for reconfiguration. Is it possible that one process doesn't catch the FINISH signal or the PID value is not the right one at scaling procedure or cluster deployment and this lead to race condition on socket acces ? May this issue come from the use of a haproxy internetSocket object instead of a unixSocket ?

**Comment by oktalz** [comment]
@Dr4s1l That is normal, HAProxy is running in master-worker mode.

**Comment by dkorunic** [comment]
@Dr4s1l That's normal. On top of what @oktalz just said about master-worker mode, you will also notice that there are additional processes on reload: then a new HAProxy process is being spawned which handles all new connections, while old process is still kept live and keeps serving old connections until they are closed as this model allows for seamless reloads.

**Comment by Dr4s1l** [comment]
Is it possible that this process lead to race condition on the unix socket?

**Comment by dkorunic** [comment]
@Dr4s1l No. If possible, it would help a lot if you can follow the guide from @Mo3m3n about logs and haproxy.cfg, that's what we need to establish where is the issue.

**Comment by Dr4s1l** [comment]
```
┌─[yggdrasil@localhost] - [~] - [2020-07-01 10:57:26]
└─[0] <> kubectl logs -n gateway haproxy-kubernetes-ingress-7b8596498f-794jh
2020/06/29 19:19:23 
 _   _    _    ____                                            
| | | |  / \  |  _ \ _ __ _____  ___   _                       
| |_| | / _ \ | |_) | '__/ _ \ \/ / | | |                      
|  _  |/ ___ \|  __/| | | (_) >  <| |_| |                      
|_| |_/_/   \_\_|   |_|  \___/_/\_\\__, |                      
 _  __     _                       |___/             ___ ____  
| |/ /   _| |__   ___ _ __ _ __   ___| |_ ___  ___  |_ _/ ___| 
| ' / | | | '_ \ / _ \ '__| '_ \ / _ \ __/ _ \/ __|  | | |      
| . \ |_| | |_) |  __/ |  | | | |  __/ ||  __/\__ \  | | |___ 
|_|\_\__,_|_.__/ \___|_|  |_| |_|\___|\__\___||___/ |___\____| 
															 

2020/06/29 19:19:23 HAProxy Ingress Controller v1.4.5 e594cad

2020/06/29 19:19:23 Build from: git@github.com:haproxytech/kubernetes-ingress.git
2020/06/29 19:19:23 Build date: 2020-06-04T18:31:12

2020/06/29 19:19:23 ConfigMap: gateway/haproxy-kubernetes-ingress
2020/06/29 19:19:23 Ingress class: haproxy
2020/06/29 19:19:23 Publish service: 
2020/06/29 19:19:23 Default backend service: gateway/haproxy-kubernetes-ingress-default-backend
2020/06/29 19:19:23 Default ssl certificate: gateway/haproxy-kubernetes-ingress-default-cert
2020/06/29 19:19:23 Controller sync period: 5s
2020/06/29 19:19:23 controller.go:254 Running with HA-Proxy version 2.1.4 2020/04/02 - https://haproxy.org/
2020/06/29 19:19:23 INFO    controller.go:259 Starting HAProxy with /etc/haproxy/haproxy.cfg
2020/06/29 19:19:23 INFO    controller.go:264 Running on haproxy-kubernetes-ingress-7b8596498f-794jh
2020/06/29 19:19:23 INFO    controller.go:95 Running on Kubernetes version: v1.18.4 linux/amd64
[NOTICE] 180/191923 (29) : New worker #1 (30) forked
2020/06/29 19:19:29 INFO    controller.go:213 HAProxy reloaded
2020/06/30 06:54:20 INFO    controller.go:213 HAProxy reloaded
2020/06/30 07:11:22 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'backend-fixtures-user-with-another-unit-nodejs' not found
2020/06/30 07:11:22 INFO    controller.go:213 HAProxy reloaded
2020/06/30 07:11:29 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'backend-fixtures-user-with-another-unit-nodejs' not found
2020/06/30 07:11:29 ERROR   controller.go:145 ingress servicePort(Str: , Int: 8089) for serviceName 'cm-acme-http-solver-zr288' not found
2020/06/30 07:11:29 INFO    controller.go:213 HAProxy reloaded
2020/06/30 07:11:38 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'backend-fixtures-user-with-another-unit-nodejs' not found
2020/06/30 07:11:43 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'backend-fixtures-user-with-another-unit-nodejs' not found
2020/06/30 07:20:49 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'frontend-fixes-roles-and-workflow-nodejs' not found
2020/06/30 07:20:49 INFO    controller.go:213 HAProxy reloaded
2020/06/30 07:20:55 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'frontend-fixes-roles-and-workflow-nodejs' not found
2020/06/30 07:21:00 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'frontend-fixes-roles-and-workflow-nodejs' not found
2020/06/30 07:21:05 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'frontend-fixes-roles-and-workflow-nodejs' not found
2020/06/30 07:21:10 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'frontend-fixes-roles-and-workflow-nodejs' not found
2020/06/30 07:21:19 ERROR   controller.go:145 ingress servicePort(Str: , Int: 8089) for serviceName 'cm-acme-http-solver-zr288' not found
2020/06/30 07:21:20 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'frontend-fixes-roles-and-workflow-nodejs' not found
2020/06/30 07:21:20 INFO    controller.go:213 HAProxy reloaded
2020/06/30 07:21:27 ERROR   controller.go:145 ingress servicePort(Str: , Int: 80) for serviceName 'frontend-fixes-roles-and-workflow-nodejs' not found
2020/06/30 07:21:44 ERROR   controller.go:145 ingress servicePort(Str: , Int: 8089) for serviceName 'cm-acme-http-solver-gv4c9' not found
2020/06/30 07:21:44 INFO    controller.go:213 HAProxy reloaded
2020/06/30 07:25:17 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80/SRV_WsTOX addr 100.65.99.216]
2020/06/30 07:25:17 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80/SRV_WsTOX state ready]
2020/06/30 07:25:18 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80/SRV_WsTOX addr 100.65.99.216]
2020/06/30 07:25:18 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80/SRV_WsTOX state ready]
2020/06/30 07:25:18 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80/SRV_DhVJ8 addr 127.0.0.1]
2020/06/30 07:25:18 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80/SRV_DhVJ8 state maint]
2020/06/30 13:17:57 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80/SRV_meF6E addr 100.65.99.225]
2020/06/30 13:17:57 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80/SRV_meF6E state ready]
2020/06/30 13:17:58 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80/SRV_meF6E addr 100.65.99.225]
2020/06/30 13:17:58 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80/SRV_meF6E state ready]
2020/06/30 13:17:58 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80/SRV_PZwyx addr 127.0.0.1]
2020/06/30 13:17:58 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80/SRV_PZwyx state maint]
2020/06/30 22:11:32 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:12:44 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:13:21 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-staging-backend-nodejs-80/SRV_4XbW6 addr 100.65.99.242]
2020/06/30 22:13:21 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-staging-backend-nodejs-80/SRV_4XbW6 state ready]
2020/06/30 22:13:44 ERROR   events.go:368 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-staging-backend-nodejs-80/SRV_4XbW6 addr 127.0.0.1]
2020/06/30 22:13:44 ERROR   events.go:378 /var/run/haproxy-runtime-api.sock [3]  No such server.
 [set server stargate-staging-backend-nodejs-80/SRV_4XbW6 state maint]
2020/06/30 22:15:00 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:20:40 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:23:56 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:33:53 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:40:53 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:54:13 INFO    controller.go:213 HAProxy reloaded
2020/06/30 22:56:28 INFO    controller.go:213 HAProxy reloaded
2020/06/30 23:05:35 INFO    controller.go:213 HAProxy reloaded
2020/07/01 07:35:44 INFO    controller.go:213 HAProxy reloaded
```
```
# _version=44
# HAProxy Technologies
# https://www.haproxy.com/
# this file is not meant to be changed directly
# it is under haproxy ingress controller management

global 
  daemon
  master-worker
  pidfile /var/run/haproxy.pid
  stats socket /var/run/haproxy-runtime-api.sock level admin expose-fd listeners
  stats timeout 1m
  tune.ssl.default-dh-param 2048
  ssl-default-bind-options no-sslv3 no-tls-tickets no-tlsv10
  ssl-default-bind-ciphers ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!3DES:!MD5:!PSK
  log 127.0.0.1:514 local0 notice
  server-state-file global
  server-state-base /var/state/haproxy/

defaults 
  log global
  log-format '%ci:%cp [%tr] %ft %b/%s %TR/%Tw/%Tc/%Tr/%Ta %ST %B %CC %CS %tsc %ac/%fc/%bc/%sc/%rc %sq/%bq %hr %hs "%HM %[var(txn.base)] %HV"'
  option redispatch
  option dontlognull
  option http-keep-alive
  timeout http-request 5s
  timeout connect 5s
  timeout client 50s
  timeout queue 5s
  timeout server 50s
  timeout tunnel 1h
  timeout http-keep-alive 1m
  load-server-state-from-file global

frontend healthz 
  mode http
  bind 0.0.0.0:1042 name healtz_1
  monitor-uri /healthz
  option dontlog-normal

frontend http 
  mode http
  bind 0.0.0.0:80 name bind_1
  bind :::80 v4v6 name bind_2
  http-request set-var(txn.base) base
  http-request redirect scheme https code 302 if { req.hdr(host),field(1,:) -f /etc/haproxy/maps/16657147599902430202.lst } !{ ssl_fc }
  use_backend stargate-staging-frontend-nodejs-80 if { req.hdr(host),field(1,:) -i stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-backend-treated-request-nodejs-80 if { req.hdr(host),field(1,:) -i review-treated-re-95c9qn.api.stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80 if { req.hdr(host),field(1,:) -i review-fixtures-u-6mmp5y.api.stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80 if { req.hdr(host),field(1,:) -i review-fixes-role-545ax3.stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-frontend-feature-pagination-nodejs-80 if { req.hdr(host),field(1,:) -i review-feature-pa-yjnrta.stargate.igloo.fabnum.fr } 
  use_backend echauffeur-reviews-api-feat-add-steps-to-shuttles-nodejs-80 if { req.hdr(host),field(1,:) -i review-feat-add-s-2p2jc1.api.echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-staging-user-nodejs-80 if { req.hdr(host),field(1,:) -i echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-reviews-driver-nodejs-80 if { req.hdr(host),field(1,:) -i driver.echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-reviews-dashboard-nodejs-80 if { req.hdr(host),field(1,:) -i dashboard.echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-staging-api-nodejs-80 if { req.hdr(host),field(1,:) -i api.echauffeur.igloo.fabnum.fr } 
  default_backend gateway-haproxy-kubernetes-ingress-default-backend-8080

frontend https 
  mode http
  bind 0.0.0.0:443 name bind_1 crt /etc/haproxy/certs ssl alpn h2,http/1.1
  bind :::443 name bind_2 crt /etc/haproxy/certs ssl v4v6 alpn h2,http/1.1
  http-request set-var(txn.base) base
  http-request set-header X-Forwarded-Proto https if { ssl_fc }
  use_backend stargate-staging-frontend-nodejs-80 if { req.hdr(host),field(1,:) -i stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-backend-treated-request-nodejs-80 if { req.hdr(host),field(1,:) -i review-treated-re-95c9qn.api.stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-backend-fixtures-user-with-another-unit-nodejs-80 if { req.hdr(host),field(1,:) -i review-fixtures-u-6mmp5y.api.stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-frontend-fixes-roles-and-workflow-nodejs-80 if { req.hdr(host),field(1,:) -i review-fixes-role-545ax3.stargate.igloo.fabnum.fr } 
  use_backend stargate-reviews-frontend-feature-pagination-nodejs-80 if { req.hdr(host),field(1,:) -i review-feature-pa-yjnrta.stargate.igloo.fabnum.fr } 
  use_backend echauffeur-reviews-api-feat-add-steps-to-shuttles-nodejs-80 if { req.hdr(host),field(1,:) -i review-feat-add-s-2p2jc1.api.echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-staging-user-nodejs-80 if { req.hdr(host),field(1,:) -i echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-reviews-driver-nodejs-80 if { req.hdr(host),field(1,:) -i driver.echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-reviews-dashboard-nodejs-80 if { req.hdr(host),field(1,:) -i dashboard.echauffeur.igloo.fabnum.fr } 
  use_backend echauffeur-staging-api-nodejs-80 if { req.hdr(host),field(1,:) -i api.echauffeur.igloo.fabnum.fr } 
  default_backend gateway-haproxy-kubernetes-ingress-default-backend-8080

frontend stats 
  mode http
  bind *:1024
  option http-use-htx
  stats enable
  stats uri /
  stats refresh 10s
  http-request use-service prometheus-exporter if { path /metrics }

backend echauffeur-reviews-api-feat-add-steps-to-shuttles-nodejs-80 
  mode http
  balance roundrobin
  option forwardfor
  server SRV_VRupg 127.0.0.1:1337 disabled check weight 128
  server SRV_cYNDy 127.0.0.1:1337 disabled check weight 128
  server SRV_FlQ8u 127.0.0.1:1337 disabled check weight 128
  server SRV_AHvtq 127.0.0.1:1337 disabled check weight 128
  server SRV_kTrcG 127.0.0.1:1337 disabled check weight 128
  server SRV_lSnqR 127.0.0.1:1337 disabled check weight 128
  server SRV_Ejzfa 127.0.0.1:1337 disabled check weight 128
  server SRV_Yt4PP 127.0.0.1:1337 disabled check weight 128
  server SRV_hfQ5K 127.0.0.1:1337 disabled check weight 128
  server SRV_McjOs 127.0.0.1:1337 disabled check weight 128
  server SRV_ZXwU8 127.0.0.1:1337 disabled check weight 128
```

**Comment by paovitali** [comment]
Hi, as reported in slack channel, I'm experiencing a similar issue. When the controller is up and running, newly created apps (process is create namespace/service/ingress then deploy via spinnaker) won't have any active backend in running memory config, until a reload is triggered by modifying a k8s object (wether it is ingress or service)

My Setup:

* EKS cluster (v1.16.8)
* HAProxy installed via chart (chart: v1.4.0 / app: 1.4.5)

The process:

1.  Haproxy is up and running (tested with 2 or 4 pods config)
2. Create K8S objects (namespace / service / ingress etc...) - nothing deployed inside the NS at this point in time
3. Deploy a canary version of the app (1st running pod)
4. Deploy a stable version of the app (1 to n running pods)

Nor after step `3` or `4` any pod shows up in the haproxy stats web page or socat server /stats details (checked inside running pod). Instead we have the proper configuration in haproxy.cfg files inside the controller pods.

from logs after step 4

```
Proxy appname-service-8081 started.
Stopping backend appname-service-8081 in 0 ms.
[WARNING] 183/140437 (380) : Stopping backend appname-service-8081 in 0 ms.
[WARNING] 183/140437 (380) : Proxy appname-service-8081 stopped (FE: 0 conns, BE: 0 conns).
Proxy appname-service-8081 stopped (FE: 0 conns, BE: 0 conns).
Proxy appname-service-8081 started.
Stopping backend appname-service-8081 in 0 ms.
[WARNING] 183/140718 (453) : Stopping backend appname-service-8081 in 0 ms.
Proxy appname-service-8081 stopped (FE: 0 conns, BE: 0 conns).
[WARNING] 183/140718 (453) : Proxy appname-service-8081 stopped (FE: 0 conns, BE: 0 conns).
```

haproxy.cfg
```
backend appname-service-8081
  mode http
  balance leastconn
  server SRV_MYusP 127.0.0.1:8081 disabled check weight 128
  server SRV_I89SI 127.0.0.1:8081 disabled check weight 128
  server SRV_YE0Z4 127.0.0.1:8081 disabled check weight 128
  server SRV_BkfPG 127.0.0.1:8081 disabled check weight 128
  server SRV_yeBAF 127.0.0.1:8081 disabled check weight 128
  server SRV_rQHJZ 127.0.0.1:8081 disabled check weight 128
  server SRV_cyLfL 127.0.0.1:8081 disabled check weight 128
  server SRV_9sXSn 127.0.0.1:8081 disabled check weight 128
  server SRV_95Zo9 127.0.0.1:8081 disabled check weight 128
  server SRV_WGzL6 127.0.0.1:8081 disabled check weight 128
  server SRV_ba9xn 127.0.0.1:8081 disabled check weight 128
  server SRV_xixrc 127.0.0.1:8081 disabled check weight 128
  server SRV_Bsq7X 127.0.0.1:8081 disabled check weight 128
  server SRV_z15XD 127.0.0.1:8081 disabled check weight 128
  server SRV_belIE 127.0.0.1:8081 disabled check weight 128
  server SRV_BnEh6 127.0.0.1:8081 disabled check weight 128
  server SRV_5RL3T 127.0.0.1:8081 disabled check weight 128
  server SRV_FtrRS 127.0.0.1:8081 disabled check weight 128
  server SRV_nr6Tm 127.0.0.1:8081 disabled check weight 128
  server SRV_eh3eZ 127.0.0.1:8081 disabled check weight 128
  server SRV_zMJkh 127.0.0.1:8081 disabled check weight 128
  server SRV_Emc2n 127.0.0.1:8081 disabled check weight 128
  server SRV_VKKIi 127.0.0.1:8081 disabled check weight 128
  server SRV_GPB9Q 127.0.0.1:8081 disabled check weight 128
  server SRV_QCqdZ 127.0.0.1:8081 disabled check weight 128
  server SRV_LWrFJ 127.0.0.1:8081 disabled check weight 128
  server SRV_BU5OS 127.0.0.1:8081 disabled check weight 128
  server SRV_o4kxA 127.0.0.1:8081 disabled check weight 128
  server SRV_kyBjX 127.0.0.1:8081 disabled check weight 128
  server SRV_piUCg 127.0.0.1:8081 disabled check weight 128
  server SRV_0i5aY 127.0.0.1:8081 disabled check weight 128
  server SRV_AlY4D 127.0.0.1:8081 disabled check weight 128
  server SRV_Dk2pO 127.0.0.1:8081 disabled check weight 128
  server SRV_sXLSx 127.0.0.1:8081 disabled check weight 128
  server SRV_mPaC4 127.0.0.1:8081 disabled check weight 128
  server SRV_2quc9 127.0.0.1:8081 disabled check weight 128
  server SRV_gubb3 127.0.0.1:8081 disabled check weight 128
  server SRV_d3Zn9 127.0.0.1:8081 disabled check weight 128
  server SRV_k57p1 127.0.0.1:8081 disabled check weight 128
  server SRV_BxEE2 127.0.0.1:8081 disabled check weight 128
  server SRV_Y6PsJ 127.0.0.1:8081 disabled check weight 128
  server SRV_vwCxZ 10.207.36.138:8081 check weight 128
```

Then to fix:
5. Modify something in the ingress / service manifests (adding an host or modifying a svc port)

```
2020/07/02 14:16:32 INFO    controller.go:213 HAProxy reloaded``
```
6. Backends are now properly picked up @ runtime by the controller pods

Unfortunately adding debug log didn't track anything useful to understand what happens inside the controller, thanks for sorting this out

**Comment by Mo3m3n** [comment]
I think I have found a bug that is quite related to what you are describing and fixed here https://github.com/haproxytech/kubernetes-ingress/commit/fede729a05aef0f3d550274da383059695d55d05
You can already test this without having to wait for the next release via this tag: `haproxytech/kubernetes-ingress:dev` and let me know your feedback

**Comment by paovitali** [comment]
@Mo3m3n Thanks for your work, I confirm you the fixed version (dev tag) works perfectly and we're not experiencing anymore the backend issues when deploying a new app, thanks for your debug and fix 👍

**Comment by lseelenbinder** [comment]
Has this fix been deployed to production? I had a similar bug yesterday (resolved by deleting the controller pods).

**Comment by Mo3m3n** [comment]
Hi this is available in `haproxytech/kubernetes-ingress:1.4.6`
Let me know if everything is ok

**Comment by Nainterceptor** [comment]
Hello, I've tested some flow with ingress 1.4.6 and it seems to be fixed. I think we can close this issue, if further tests have opposite results, I'll reopen.
